exercises/advanced/caching-deep-dive.md

```diff
@@ -0,0 +1,115 @@
+# `📖 Excercise:` Cache miss debugging, optimizations and fine tuning of inputs and outputs
+
+## 📚&nbsp;&nbsp;**Learning outcomes**
+- Learn how to debug cache misses
+- Understand how task inputs work and how to configure them for the best caching results
+
+## 🏋️‍♀️&nbsp;&nbsp;Steps:
+
+### 1. Understanding the task hash
+
+Everytime we run a task, Nx will associate a hash signature to that task run.
+
+The hash signature is based on following variables:
+- Command being run
+- Project's configuration (and its dependencies)
+- Contents of task inputs (and its dependencies)
+- Executor being invoked by the command
+
+First parameter is obvious. If you run `nx build movies-app`, Nx will store the command signature in the hash details and use it to calculate the final hash. Every time the command changes ever so slightly, for example when running it with `--prod` or `--verbose`, will change the final hash stamp.
+
+Each tasks is defined by list of inputs that define the context upon which we run the task command. Inputs can be defined in `nx.json` as `targetDefaults`, in `project.json` and `package.json` target's configuration or any combination of the above. The best way to get the full list of inputs is by using `Nx Console` IDE extension or by running `nx graph`.
+
+![Lint task inputs](../assets/lint-inputs.png)
+
+Some files are explicitly specified, while others are represented by `namedInputs`. Named inputs are variables that encapsulate common file groups. The two typical ones are:
+- `default` - includes all the files in the project
+- `production` - exludes all files that are not production relevant like `spec.ts` files and lint and test configurations.
+
+Any change to files included in the inputs will result in the new hash signature.
+
+Try running the `build` for `movies-api`. Running the build for the second time should result in the cache hit. If you change for example `main.ts` next run will result in cache miss. But if we change the `app.controller.spec.ts` our `build` should still be retrieved from the cache, because we didn't change the production file set.
```

Comment:
should we have them change the namedInputs in a way s.t. it excludes tests. and then see the effect of it? or some other example. I often use the exluding of markdown files from builds etc.

Reply:
For the simplicity let's give them this task verbally. But I agree it makes sense.